stress: failed test in cockroach/gossip/gossip.test: TestGossipCullNetwork

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/ee73b1d854981df7e7841466492f52252f2dacaf

Stress build found a failed test:

```
=== RUN   TestGossipCullNetwork
W160713 05:31:50.808749 gossip/gossip.go:934  not connected to cluster; use --join to specify a connected node
W160713 05:31:50.810334 gossip/gossip.go:934  not connected to cluster; use --join to specify a connected node
I160713 05:31:50.812738 gossip/gossip.go:960  starting client to 127.0.0.1:36985
W160713 05:31:50.812833 gossip/gossip.go:934  not connected to cluster; use --join to specify a connected node
I160713 05:31:50.817952 gossip/gossip.go:960  starting client to 127.0.0.1:33525
W160713 05:31:50.818053 gossip/gossip.go:934  not connected to cluster; use --join to specify a connected node
I160713 05:31:50.819533 gossip/gossip.go:960  starting client to 127.0.0.1:43814
W160713 05:31:50.837268 gossip/gossip.go:938  partition in gossip network; attempting new connection
W160713 05:31:50.837295 gossip/gossip.go:938  partition in gossip network; attempting new connection
--- FAIL: TestGossipCullNetwork (15.11s)
	testing.go:113: condition failed to evaluate within 15s: no network culling occurred
```

Run Details:

```
83 runs so far, 0 failures, over 5s
182 runs so far, 0 failures, over 10s
276 runs so far, 0 failures, over 15s
371 runs so far, 0 failures, over 20s
461 runs so far, 0 failures, over 25s
549 runs so far, 0 failures, over 30s
598 runs completed, 1 failures, over 33s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/gossip/client.go
+++ b/gossip/client.go
@@ -68,6 +68,8 @@ func newClient(addr net.Addr) *client {
 // the client is sent on the disconnected channel. This method starts client
 // processing in a goroutine and returns immediately.
 func (c *client) start(g *Gossip, disconnected chan *client, ctx *rpc.Context, stopper *stop.Stopper) {
+	log.Infof("starting client to %s", c.addr)
+
 	stopper.RunWorker(func() {
 		defer func() {
 			disconnected <- c

--- a/gossip/client_test.go
+++ b/gossip/client_test.go
@@ -223,7 +223,7 @@ func TestClientDisconnectLoopback(t *testing.T) {
 	// startClient requires locks are held, so acquire here.
 	local.mu.Lock()
 	lAddr := local.is.NodeAddr
-	local.startClient(&lAddr, stopper)
+	local.startClient(&lAddr)
 	local.mu.Unlock()
 	local.manage()
 	util.SucceedsSoon(t, func() error {
@@ -250,8 +250,8 @@ func TestClientDisconnectRedundant(t *testing.T) {
 
 	rAddr := remote.is.NodeAddr
 	lAddr := local.is.NodeAddr
-	local.startClient(&rAddr, stopper)
-	remote.startClient(&lAddr, stopper)
+	local.startClient(&rAddr)
+	remote.startClient(&lAddr)
 	local.mu.Unlock()
 	remote.mu.Unlock()
 	local.manage()
@@ -289,8 +289,8 @@ func TestClientDisallowMultipleConns(t *testing.T) {
 	// Start two clients from local to remote. RPC client cache is
 	// disabled via the context, so we'll start two different outgoing
 	// connections.
-	local.startClient(&rAddr, stopper)
-	local.startClient(&rAddr, stopper)
+	local.startClient(&rAddr)
+	local.startClient(&rAddr)
 	local.mu.Unlock()
 	remote.mu.Unlock()
 	local.manage()

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -870,23 +870,25 @@ func (g *Gossip) tightenNetwork(distantNodeID roachpb.NodeID) {
 // outgoing node to free up an outgoing spot for a more targeted
 // tightening (via tightenNetwork).
 func (g *Gossip) cullNetwork() {
-	g.mu.Lock()
-	defer g.mu.Unlock()
-	// If there's no space, find and remove least useful peer, if possible.
-	if g.outgoing.hasSpace() {
-		return
-	}
-	leastUsefulID := g.is.leastUseful(g.outgoing)
+	leastUsefulID := func() roachpb.NodeID {
+		g.mu.Lock()
+		defer g.mu.Unlock()
+		// If there's no space, find and remove least useful peer, if possible.
+		if g.outgoing.hasSpace() {
+			return 0
+		}
+		return g.is.leastUseful(g.outgoing)
+	}()
 	if leastUsefulID == 0 {
 		if log.V(1) {
 			log.Infof("couldn't find least useful client to close")
 		}
-		return
+	} else {
+		if log.V(1) {
+			log.Infof("closing least useful client to node %d to tighten network graph", leastUsefulID)
+		}
+		g.closeClient(leastUsefulID)
 	}
-	if log.V(1) {
-		log.Infof("closing least useful client to node %d to tighten network graph", leastUsefulID)
-	}
-	g.closeClient(leastUsefulID)
 }
 
 func (g *Gossip) doDisconnected(c *client) {
@@ -964,6 +966,7 @@ func (g *Gossip) startClient(addr net.Addr) {
 func (g *Gossip) closeClient(nodeID roachpb.NodeID) {
 	if c := g.findClient(func(c *client) bool { return c.peerID == nodeID }); c != nil {
 		c.close()
+		g.doDisconnected(<-g.disconnected) // block until the client has disconnected
 	}
 }
 

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -766,13 +766,11 @@ func (g *Gossip) getNextBootstrapAddress() net.Addr {
 // receives notifications that gossip network connectivity has been
 // lost and requires re-bootstrapping.
 func (g *Gossip) bootstrap() {
-	stopper := g.server.stopper
-
-	stopper.RunWorker(func() {
+	g.server.stopper.RunWorker(func() {
 		var bootstrapTimer timeutil.Timer
 		defer bootstrapTimer.Stop()
 		for {
-			if stopper.RunTask(func() {
+			if g.server.stopper.RunTask(func() {
 				g.mu.Lock()
 				defer g.mu.Unlock()
 				haveClients := g.outgoing.len() > 0
@@ -780,7 +778,7 @@ func (g *Gossip) bootstrap() {
 				if !haveClients || !haveSentinel {
 					// Try to get another bootstrap address from the resolvers.
 					if addr := g.getNextBootstrapAddress(); addr != nil {
-						g.startClient(addr, stopper)
+						g.startClient(addr)
 					} else {
 						// We couldn't start a client, signal that we're stalled so that
 						// we'll retry.
@@ -797,14 +795,14 @@ func (g *Gossip) bootstrap() {
 			case <-bootstrapTimer.C:
 				bootstrapTimer.Read = true
 				// break
-			case <-stopper.ShouldStop():
+			case <-g.server.stopper.ShouldStop():
 				return
 			}
 			// Block until we need bootstrapping again.
 			select {
 			case <-g.stalled:
 				// break
-			case <-stopper.ShouldStop():
+			case <-g.server.stopper.ShouldStop():
 				return
 			}
 		}
@@ -822,21 +820,19 @@ func (g *Gossip) bootstrap() {
 // connections or the sentinel gossip is unavailable, the bootstrapper
 // is notified via the stalled conditional variable.
 func (g *Gossip) manage() {
-	stopper := g.server.stopper
-
-	stopper.RunWorker(func() {
+	g.server.stopper.RunWorker(func() {
 		cullTicker := time.NewTicker(g.jitteredInterval(g.cullInterval))
 		stallTicker := time.NewTicker(g.jitteredInterval(g.stallInterval))
 		defer cullTicker.Stop()
 		defer stallTicker.Stop()
 		for {
 			select {
-			case <-stopper.ShouldStop():
+			case <-g.server.stopper.ShouldStop():
 				return
 			case c := <-g.disconnected:
-				g.doDisconnected(stopper, c)
+				g.doDisconnected(c)
 			case nodeID := <-g.tighten:
-				g.tightenNetwork(stopper, nodeID)
+				g.tightenNetwork(nodeID)
 			case <-cullTicker.C:
 				g.cullNetwork()
 			case <-stallTicker.C:
@@ -857,7 +853,7 @@ func (g *Gossip) jitteredInterval(interval time.Duration) time.Duration {
 // tightenNetwork "tightens" the network by starting a new gossip
 // client to the most distant node as measured in required gossip hops
 // to propagate info from the distant node to this node.
-func (g *Gossip) tightenNetwork(stopper *stop.Stopper, distantNodeID roachpb.NodeID) {
+func (g *Gossip) tightenNetwork(distantNodeID roachpb.NodeID) {
 	g.mu.Lock()
 	defer g.mu.Unlock()
 	if g.outgoing.hasSpace() {
@@ -865,7 +861,7 @@ func (g *Gossip) tightenNetwork(stopper *stop.Stopper, distantNodeID roachpb.Nod
 			log.Errorf("node %d: %s", distantNodeID, err)
 		} else {
 			log.Infof("starting client to distant node %d to tighten network graph", distantNodeID)
-			g.startClient(nodeAddr, stopper)
+			g.startClient(nodeAddr)
 		}
 	}
 }
@@ -893,14 +889,14 @@ func (g *Gossip) cullNetwork() {
 	g.closeClient(leastUsefulID)
 }
 
-func (g *Gossip) doDisconnected(stopper *stop.Stopper, c *client) {
+func (g *Gossip) doDisconnected(c *client) {
 	g.mu.Lock()
 	defer g.mu.Unlock()
 	g.removeClient(c)
 
 	// If the client was disconnected with a forwarding address, connect now.
 	if c.forwardAddr != nil {
-		g.startClient(c.forwardAddr, stopper)
+		g.startClient(c.forwardAddr)
 	}
 	g.maybeSignalStalledLocked()
 }
@@ -956,12 +952,12 @@ func (g *Gossip) checkHasConnected() {
 // startClient launches a new client connected to remote address.
 // The client is added to the outgoing address set and launched in
 // a goroutine.
-func (g *Gossip) startClient(addr net.Addr, stopper *stop.Stopper) {
+func (g *Gossip) startClient(addr net.Addr) {
 	c := newClient(addr)
 	g.clientsMu.Lock()
 	g.clients = append(g.clients, c)
 	g.clientsMu.Unlock()
-	c.start(g, g.disconnected, g.rpcContext, stopper)
+	c.start(g, g.disconnected, g.rpcContext, g.server.stopper)
 }
 
 // closeClient finds and closes a client.

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -834,7 +834,31 @@ func (g *Gossip) manage() {
 			case nodeID := <-g.tighten:
 				g.tightenNetwork(nodeID)
 			case <-cullTicker.C:
-				g.cullNetwork()
+				func() {
+					g.mu.Lock()
+					if !g.outgoing.hasSpace() {
+						leastUsefulID := g.is.leastUseful(g.outgoing)
+
+						if c := g.findClient(func(c *client) bool {
+							return c.peerID == leastUsefulID
+						}); c != nil {
+							if log.V(1) {
+								log.Infof("closing least useful client %+v to tighten network graph", c)
+							}
+							c.close()
+
+							// After releasing the lock, block until the client disconnects.
+							defer func() {
+								g.doDisconnected(<-g.disconnected)
+							}()
+						} else {
+							if log.V(1) {
+								log.Infof("couldn't find least useful client among %+v", g.clients)
+							}
+						}
+					}
+					g.mu.Unlock()
+				}()
 			case <-stallTicker.C:
 				g.mu.Lock()
 				g.maybeSignalStalledLocked()
@@ -863,31 +887,6 @@ func (g *Gossip) tightenNetwork(distantNodeID roachpb.NodeID) {
 			log.Infof("starting client to distant node %d to tighten network graph", distantNodeID)
 			g.startClient(nodeAddr)
 		}
-	}
-}
-
-// cullNetwork is called periodically to remove the least "useful"
-// outgoing node to free up an outgoing spot for a more targeted
-// tightening (via tightenNetwork).
-func (g *Gossip) cullNetwork() {
-	leastUsefulID := func() roachpb.NodeID {
-		g.mu.Lock()
-		defer g.mu.Unlock()
-		// If there's no space, find and remove least useful peer, if possible.
-		if g.outgoing.hasSpace() {
-			return 0
-		}
-		return g.is.leastUseful(g.outgoing)
-	}()
-	if leastUsefulID == 0 {
-		if log.V(1) {
-			log.Infof("couldn't find least useful client to close")
-		}
-	} else {
-		if log.V(1) {
-			log.Infof("closing least useful client to node %d to tighten network graph", leastUsefulID)
-		}
-		g.closeClient(leastUsefulID)
 	}
 }
 
@@ -960,14 +959,6 @@ func (g *Gossip) startClient(addr net.Addr) {
 	g.clients = append(g.clients, c)
 	g.clientsMu.Unlock()
 	c.start(g, g.disconnected, g.rpcContext, g.server.stopper)
-}
-
-// closeClient finds and closes a client.
-func (g *Gossip) closeClient(nodeID roachpb.NodeID) {
-	if c := g.findClient(func(c *client) bool { return c.peerID == nodeID }); c != nil {
-		c.close()
-		g.doDisconnected(<-g.disconnected) // block until the client has disconnected
-	}
 }
 
 // removeClient removes the specified client. Called when a client

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -957,7 +957,6 @@ func (g *Gossip) checkHasConnected() {
 // The client is added to the outgoing address set and launched in
 // a goroutine.
 func (g *Gossip) startClient(addr net.Addr, stopper *stop.Stopper) {
-	log.Infof("starting client to %s", addr)
 	c := newClient(addr)
 	g.clientsMu.Lock()
 	g.clients = append(g.clients, c)

--- a/gossip/gossip_test.go
+++ b/gossip/gossip_test.go
@@ -18,10 +18,10 @@ package gossip
 
 import (
 	"bytes"
-	"errors"
 	"testing"
 	"time"
 
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
 
@@ -194,19 +194,19 @@ func TestGossipCullNetwork(t *testing.T) {
 	local.mu.Unlock()
 
 	util.SucceedsSoon(t, func() error {
-		if len(local.Outgoing()) == minPeers {
-			return nil
+		if peers := len(local.Outgoing()); peers != minPeers {
+			return errors.Errorf("%d of %d peers connected", peers, minPeers)
 		}
-		return errors.New("some peers not yet connected")
+		return nil
 	})
 
 	local.manage()
 
 	util.SucceedsSoon(t, func() error {
 		// Verify that a client is closed within the cull interval.
-		if len(local.Outgoing()) == minPeers-1 {
-			return nil
+		if peers := len(local.Outgoing()); peers != minPeers-1 {
+			return errors.Errorf("%d of %d peers connected", peers, minPeers-1)
 		}
-		return errors.New("no network culling occurred")
+		return nil
 	})
 }

--- a/gossip/gossip_test.go
+++ b/gossip/gossip_test.go
@@ -189,7 +189,7 @@ func TestGossipCullNetwork(t *testing.T) {
 	local.mu.Lock()
 	for i := 0; i < minPeers; i++ {
 		peer := startGossip(roachpb.NodeID(i+2), stopper, t)
-		local.startClient(&peer.is.NodeAddr, stopper)
+		local.startClient(&peer.is.NodeAddr)
 	}
 	local.mu.Unlock()
 

--- a/gossip/gossip_test.go
+++ b/gossip/gossip_test.go
@@ -193,20 +193,26 @@ func TestGossipCullNetwork(t *testing.T) {
 	}
 	local.mu.Unlock()
 
-	util.SucceedsSoon(t, func() error {
+	const slowGossipDuration = time.Minute
+
+	if err := util.RetryForDuration(slowGossipDuration, func() error {
 		if peers := len(local.Outgoing()); peers != minPeers {
 			return errors.Errorf("%d of %d peers connected", peers, minPeers)
 		}
 		return nil
-	})
+	}); err != nil {
+		t.Fatalf("condition failed to evaluate within %s: %s", slowGossipDuration, err)
+	}
 
 	local.manage()
 
-	util.SucceedsSoon(t, func() error {
+	if err := util.RetryForDuration(slowGossipDuration, func() error {
 		// Verify that a client is closed within the cull interval.
 		if peers := len(local.Outgoing()); peers != minPeers-1 {
 			return errors.Errorf("%d of %d peers connected", peers, minPeers-1)
 		}
 		return nil
-	})
+	}); err != nil {
+		t.Fatalf("condition failed to evaluate within %s: %s", slowGossipDuration, err)
+	}
 }

